sycl/include/sycl/ext/oneapi/barrier.hpp

```diff
@@ -0,0 +1,134 @@
+//==--- barrier.hpp - SYCL_ONEAPI_BARRIER  ---------------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#define SYCL_EXT_ONEAPI_CUDA_ASYNC_BARRIER 1
+
+#include <CL/__spirv/spirv_ops.hpp>
+#include <cstddef>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl::ext::oneapi::cuda {
+
+class barrier {
+  int64_t state;
+
+public:
+  using arrival_token = int64_t;
+
+  // barriers cannot be moved or copied
+  barrier(const barrier &other) = delete;
+  barrier(barrier &&other) noexcept = delete;
+  barrier &operator=(const barrier &other) = delete;
+  barrier &operator=(barrier &&other) noexcept = delete;
+
+  void initialize(uint32_t expected_count) {
+#ifdef __SYCL_DEVICE_ONLY__
+    __clc_BarrierInitialize(&state, expected_count);
+#else
+    throw runtime_error("Barrier is not supported on host device.",
+                        PI_INVALID_DEVICE);
+#endif
```

Comment:
Similar to my comment above, it would be nice if there was a way for developers to get a meaningful error if they try to use this extension on anything except the CUDA backend.

Actually, now that I think about it I don't see anything in the extension itself saying that this is specific to the CUDA backend.  There should probably be a paragraph saying that the extension is only supported on the CUDA backend, and explaining what happens if you try to use the extension elsewhere (whether that's a compile-time warning/error, runtime error or just undefined behavior).

Reply:
Added explanation.